### PR TITLE
GAP-2011-Add-None-To-Department

### PR DESCRIPTION
**Pull request by ryan-tco**
https://github.com/cabinetoffice/gap-user-service/assets/135323857/1e1dd66b-524c-4f26-9393-0964edcd3fe7

### CODE REVIEW

File: src/test/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserServiceTest.java [hunk 1]
```diff
@@ -350,5 +350,22 @@ void updateRolesShouldAddApplicantAndFindRolesWhenNoRolesPresent() {
         assertThat(updatedUser.getRoles().size()).isEqualTo(4);
         assertThat(updatedUser.getRoles().stream().anyMatch(role -> role.getName().equals(RoleEnum.APPLICANT))).isTrue();
         assertThat(updatedUser.getRoles().stream().anyMatch(role -> role.getName().equals(RoleEnum.FIND))).isTrue();
-    }}
+    }
```
Comment (dylanwrightCO) [inline src/test/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserServiceTest.java:353]: You can write more test cases to cover department being null and user having admin/super admin role(s)

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 2]
```diff
@@ -117,6 +118,12 @@ private void addRoleIfNotPresent(User user, RoleEnum roleName) {
         }
     }
 
+    private void deleteDepartmentIfPresentAndUserIsNotAdminOrSuperAdmin(User user) {
+        if (user.getDepartment() != null && user.getRoles().stream().noneMatch(role -> role.getName().equals(RoleEnum.ADMIN) || role.getName().equals(RoleEnum.SUPER_ADMIN))) {
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:122]: this `user.getRoles().stream().noneMatch(role -> role.getName().equals(RoleEnum.ADMIN) || role.getName().equals(RoleEnum.SUPER_ADMIN))` feels a bit clunky coming from JS but not sure how else you'd do it in Java, so happy with it as-is

File: src/test/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserServiceTest.java [hunk 3]
```diff
@@ -350,5 +350,22 @@ void updateRolesShouldAddApplicantAndFindRolesWhenNoRolesPresent() {
...
+        Integer userId = 1;
+        List<Integer> newRoles = List.of(1, 2);
+        User user = User.builder().gapUserId(userId).department(Department.builder().name("test").build()).build();
+        when(userRepository.findById(userId)).thenReturn(Optional.of(user));
+        when(roleRepository.findById(1)).thenReturn(Optional.of(Role.builder().name(RoleEnum.APPLICANT).build()));
+        when(roleRepository.findById(2)).thenReturn(Optional.of(Role.builder().name(RoleEnum.FIND).build()));
+        User updatedUser = oneLoginUserService.updateRoles(1 , newRoles);
+
+        assertThat(updatedUser.getDepartment()).isNull();
+    }
```
[the remark this answers is not in the document]
Comment (ryan-tco) [reply src/test/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserServiceTest.java:366]: Added this

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 2 above)
Comment (ryan-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Only other way I can think is by extracting the 2 predicates and do an `or` comparison

**Review by dominicwest — COMMENTED** [review]
Probably also want to restrict the `updateDepartment` method in `OneLoginUserService` to only work on admin/super admin users

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 4]
```diff
@@ -117,6 +118,18 @@ private void addRoleIfNotPresent(User user, RoleEnum roleName) {
...
+           if (isUserOnlyApplicantOrFindWithADepartment(user)) {
+               user.setDepartment(null);
+            }
+    }
+
+    private boolean isUserOnlyApplicantOrFindWithADepartment(User user) {
+        return user.getRoles().size() == 2
+                && user.getDepartment() != null
+                && user.getRoles().stream().anyMatch(role -> role.getName().equals(RoleEnum.FIND) || role.getName().equals(RoleEnum.APPLICANT));
+    }
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:131]: would this fail if user had roles APPLICANT and ADMIN?

Comment (ConorFayleTCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:131]: don't think it would but could be worth adding a test for that

**Comment by ryan-tco** [comment]
> Probably also want to restrict the `updateDepartment` method in `OneLoginUserService` to only work on admin/super admin users

Update Department needs to work on Applicant and Find to null them when they're demoted from admi

**Comment by ryan-tco** [comment]
> > > Probably also want to restrict the updateDepartment method in OneLoginUserService to only work on admin/super admin users
> > 
> > 
> > Update Department needs to work on Applicant and Find to null them when they're demoted from admin
> 
> think he means the updateDepartment method - doesn't look like you're using that to null the user's department `public User updateDepartment(Integer id, Integer departmentId) {` but I'd argue it's a separate ticket since you'd want FE changes to disable the button or something as well

Restricted the use of updating department endpoint on users in latest, front end should be covered by

https://github.com/cabinetoffice/gap-find-apply-web/pull/63

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java [hunk 5]
```diff
@@ -46,6 +46,11 @@ public ResponseEntity<UserDto> getUserById(HttpServletRequest httpRequest, @Path
     @PatchMapping("/user/{userId}/department")
     public ResponseEntity<User> updateDepartment(HttpServletRequest httpRequest, @PathVariable("userId") Integer userId,
                                                    @RequestParam(value = "departmentId", required = false) Integer departmentId) {
+
+        if(oneLoginUserService.isUserApplicantAndFindOnly(oneLoginUserService.getUserById(userId))) {
+            throw new ForbiddenException();
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:51]: I think this should just be a 400 instead of a 403, since the client is permitted to perform this action, it's just not allowed for this user

Comment (ryan-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:52]: Fixed in latest commit

**Review by ConorFayleTCO — DISMISSED** [review]
happy with this but need to wait til prod is stable before release

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java [hunk 6]
```diff
@@ -46,10 +47,15 @@ public ResponseEntity<UserDto> getUserById(HttpServletRequest httpRequest, @Path
     @PatchMapping("/user/{userId}/department")
     public ResponseEntity<User> updateDepartment(HttpServletRequest httpRequest, @PathVariable("userId") Integer userId,
                                                    @RequestParam(value = "departmentId", required = false) Integer departmentId) {
+
         if (!roleService.isSuperAdmin(httpRequest)) {
             throw new ForbiddenException();
         }
 
+        if(oneLoginUserService.isUserApplicantAndFindOnly(oneLoginUserService.getUserById(userId))) {
+            throw new InvalidRequestException("User only has find and applicant roles");
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:56]: maybe a message like "Users with find and applicant roles cannot be assigned a department"

**Comment by ryan-tco** [comment]
Behaves as expected after dev merge

**Review by ConorFayleTCO — APPROVED** [review]
LGTM

**Review by dylanwrightCO — APPROVED** [review]
